### DISCUSSION TITLE
BRouter avoid cities / noise / traffic, prefer forests / rivers

### DISCUSSION THREAD

**Discussion by devemux86** [devemux86/cruiser#248] [Q&A] [answered]
Next Cruiser will have new BRouter profile options in its **driving-cruiser** profile and all other BRouter profiles:

- **Avoid cities**: bypass cities / big towns as far as possible
- **Prefer forests and green areas**: favor ways in forest and green, calm areas
- **Prefer rivers and seas**: favor ways along rivers or seas
- **Avoid noise**: consider the noise / pollution on the route
- **Avoid traffic**: consider traffic estimates

Technical details: https://github.com/abrensch/brouter/issues/486

<img src="https://github.com/devemux86/cruiser/assets/3484020/0c80b1fd-8e4a-4f46-a74d-6019052d43fb" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/a8330390-b481-483c-81ba-dd9bae04ecde" width="50%"/>

**Comment by Ciutnik** [comment]
I am very curious about new options!!! I do not know any other navigator offering such miracles 🙂

_Some time ago in the web version of BRouter I played with the "Prefer rivers and seas" option and it was strictly kayak navigation 😀 In fact, the navigation did not go along the waters on the roads, but flowed on this water 🙂_

**Reply by devemux86** [reply]
> I do not know any other navigator offering such miracles :slightly_smiling_face: 

They are definitely doing a great job in developing the BRouter routing.
The most important part is that everything works **offline** and the profiles offer endless possibilities for customization.

[BRouter website](https://brouter.de/brouter-web/) already has the new options in the Fastbike, Trekking and Hiking profiles.
I am currently adapting Cruiser's Driving profile with the new options.

**Reply by DolomitesTom** [reply]
Yes offline availability of routing is 👍

**Comment by Cer-o** [comment]
I like the new options, especially avoiding cities and towns.

**Comment by umpfmpf** [comment]
It would be perfect to have an Option for not to drive a road twice.

**Comment by umpfmpf** [comment]
And how an i make this to work?


-------- Ursprüngliche Nachricht --------
Von: Tom ***@***.***>
Gesendet: 21. Juli 2023 07:30:03 MESZ
An: devemux86/cruiser ***@***.***>
CC: umpfmpf ***@***.***>, Comment ***@***.***>
Betreff: Re: [devemux86/cruiser] BRouter avoid cities / traffic, prefer forests / rivers (Discussion #248)

Yes offline availability of routing is 👍

-- 
Reply to this email directly or view it on GitHub:
https://github.com/devemux86/cruiser/discussions/248#discussioncomment-6505869

**Reply by devemux86** [reply]
@umpfmpf it will be available in the next Cruiser update.

**Comment by devemux86** [comment]
Customizing the **driving-cruiser** profile I see that the new options we are likely to use a lot are:
- Avoid cities
- Prefer forests and green areas
- Prefer rivers and seas

They can create some very interesting routes, especially in combination with the various avoidances.

---

The other new options work globally on the whole route, meaning they are heavier, so they need more waypoints:
- Avoid noise
- Avoid traffic

This is always a hint for BRouter: **Use more waypoints**, it speeds up the route calculations.

**Comment by devemux86** [comment, marked as answer]
BRouter (offline) new options to avoid cities / noise / traffic and prefer forests / rivers are available in [Cruiser 3.0.26](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

Available in Cruiser's custom "**driving-cruiser**" profile and all other BRouter profiles.

**Comment by DolomitesTom** [comment]
The new options generate some nice routes... one need to play around a bit more. Fancy stuff !!

**Comment by devemux86** [comment]
Remember that BRouter also offers the very useful "**Ignore closed roads**",
that helps with temporary closures, especially if we plan for different season.

**Reply by MotoUKRider** [reply]
@devemux86 Has this function changed in the last couple of Betas?

I ask because I was unable to route through a `barrier=lift_gate` until I enabled "Ignore closed roads".  Oddly, the offending objects in OSM (which I've now deleted) were either side of a level-crossing with its own "half-barrier" tag and had been present for 3-years.

[This is the BRouter route link](https://brouter.de/brouter-web/#map=13/53.5313/-0.5164/standard&lonlats=-0.513525,53.547747;-0.528942,53.511345&profile=car-fast) and the diversion away from the road with the railway crossing should be obvious.

We've had a LOT of discussion about this in a separate thread (cattle-grids etc) but this was the most obvious reference to "closed roads" that I could find.  Things were OK until recently, but exactly when I am not sure.  I went back to 5.5.0.b3 but the problem persisted.  Perhaps this is a change to the internal algorithm of BRouter which Cruiser picks-up as part of the regular builds?

I cannot find the setting in a BRouter profile so haven't been able to replicate the "fix" using BRouter "Web" - except to say that the "Car fast" online profile re-routes, too. "Trekking bike" routes OK.  Desktop Cruiser, using BRouter "online" Car/Motorcycle profiles also has the re-route, but Bike is OK.

The next step for is to wait a day for the routing tile to update so I can confirm my edit.

**Reply by devemux86** [reply]
@MotoUKRider The current driving profile: [driving.brf.txt](https://github.com/user-attachments/files/20532910/driving.brf.txt)

The `ignore_barrier` option checks `barrier=gate|bollard|lift_gate|cycle_barrier`.
That's how it's been working all this time.

**Reply by MotoUKRider** [reply]
@devemux86 Thanks for the profile, I may keep that "on file" :smile: 

Loading that profile in the online panel works fine, I change the `ignore_barrier` to "false" and it diverts like in Cruiser App. So I've pin-pointed the setting at work.

What I suppose I'm seeing (and saying) is that `Ignore closed roads` (in Cruiser) and `ignore_barrier` (BRouter) are the same setting?  I was under the impression that a "closed road" has its own value in OSM, but appreciate a physical object would also prevent travel for some types of vehicle.

I can't understand why a barrier object, seemingly present for 3-years, has suddenly become an issue.  What I think I'll do, is look for other level-crossings to see if there are any separate "barriers", perhaps as a result of a poor-edit.  I guess that's the problem, but I am *sure* everything was working OK a short time ago :confused:

**Reply by devemux86** [reply]
> What I suppose I'm seeing (and saying) is that Ignore closed roads (in Cruiser) and ignore_barrier (BRouter) are the same setting?

@MotoUKRider This is correct.

**Comment by Leeb64** [comment]
Hmm...(nagging mode)
Depending on the setting, the new BRouter profile generates partly usable routes.
Bypassing the cities doesn't work at all, no matter what I set (avoid cities, don't avoid main roads, avoid residential streets, high speed + high weight). Although there is a bypass road in several cases, the route leads through the middle of the town.
The function with "prefer forest and green regions" can still be increased. In some cases, it routes directly in the opposite direction-> out of the forest into the plain.
I would also find it very useful if a direct link to BRouter profile would be added under Options/Advanced.
So you can select your own BRouter profile relatively quickly.

**Reply by devemux86** [reply]
> I would also find it very useful if a direct link to BRouter profile would be added under Options/Advanced.

It is already available in "Brouter (offline) | Advanced", as most users will not know or even use it.

**Reply by devemux86** [reply]
@Leeb64 thanks for the feedback!

Do you have some specific examples so we can test?
Please also mention the BRouter options.

It really depends on the enabled options and the specific area and road network.
Some options may fight each other and too many enabled options fun begins. :slightly_smiling_face:

**Comment by devemux86** [comment]
I think they mean cities, not small towns or villages.
(the technical details are linked in the 1st post, if someone wants to search how it works)

Here is the custom **driving-cruiser** profile, you can also load it in [BRouter website](https://brouter.de/brouter-web/) for route planning.

[driving-cruiser.brf.zip](https://github.com/devemux86/cruiser/files/12136207/driving-cruiser.brf.zip)

And then export the route as gpx navigation track and load it offline in Cruiser, see instructions here:
- https://github.com/devemux86/cruiser/discussions/250#discussioncomment-6508754

**Reply by Leeb64** [reply]
I mean all sorts of villages around which a bypass road leads.
It's annoying to drive through a town when a road leads around it....
The thing with the forests isn't really effective either, ->mountains would be better(is there a possibility?).
With us in the vicinity is the largest contiguous forest area in Germany. Totally flat with dead straight roads. More or less boring there(with the Motocycle).The BRouter tries to guide me there.....

**Reply by devemux86** [reply]
Copying from their technical discussion:
https://github.com/abrensch/brouter/issues/486#issuecomment-1378441502

_"2- new tag "town"
This tag is intended for bikers who like to bypass cities and big towns.
Each highway within the administrative surface of the city/town get a tag "estimated_town_class" between 1 and 6.
Exceptions:
Highways having a tag such as "estimated_river_class" or "estimated_forest_class" are not considered, this eliminated areas of the town which are in forests or at river sides."_

---

It is not to avoid all residential areas, only cities, not villages.
Maybe the "Avoid noise" or "Avoid residential roads" (highway=residential|service|living_street) can also help.

We need a lot of testing to understand all BRouter options and which ones can be best combined together.

**Comment by devemux86** [comment]
In the [BRouter topic](https://github.com/abrensch/brouter/issues/486#issuecomment-1646872290) we can see how they categorize the towns based on the population:

_population > '2000000' then class 6
population > '1000000' then class 5
population > '400000' then class 4
population > '150000' then class 3
population > '80000' then class 2
population > '50000' then class 1_

---

Cruiser currently does not make exceptions in "Avoid towns", it declares to avoid all the above categories.

**Comment by Leeb64** [comment]
Well... after a few tries with the options: prefer forests+green regions, prefer rivers+seas, avoid noise, avoid traffic - I have to say: That's not good for much.
It often has the same effect with all settings.Often(most) the results are not really good.Sometimes nothing is calculated (waypoint too far from the route (WP is on a road)).The calculation time is too long with this options...
It doesn't really help at all.
The Kurviger algorithm is light years better. The rest of Kurviger is light years worse...
Can that be done with the BRouter options???
I think mountainous regions, gradients would be an approach.
In addition, avoiding and driving around towns.
I made a profile for myself.
It is an ECO profile (in principle) which bypasses localities, does not avoid inclines. You should set a WP about every 30-80km.

[__MOPPED__ (+) .brf.txt](https://github.com/devemux86/cruiser/files/12146841/__MOPPED__.%2B.brf.txt)

**Reply by devemux86** [reply]
@Leeb64 thanks for the feedback!

I am curious, have you tested the other BRouter profiles in Cruiser: Bike or Hiking?
These are the original BRouter profiles, so wondering if they have similar behavior?

In such case, it would be useful to report this feedback also to BRouter developers, so they know what users have found:
(they still improve things)
- https://github.com/abrensch/brouter/issues/486

**Comment by DolomitesTom** [comment]
> population > '2000000' then class 6
population > '1000000' then class 5
population > '400000' then class 4
population > '150000' then class 3
population > '80000' then class 2
population > '50000' then class 1

I find the new options quite useful. But it leads to routing through small villages instead of staying on the main road even if main roads are not excluded. This is not that good. 
Example: this is something I don't like at all, instead of staying on the main road it routes through smallest streets  within a village with a few houses, although avoiding residential streets is enabled:
![Screenshot_20230725-081805_Cruiser](https://github.com/devemux86/cruiser/assets/121371629/96ca25c0-0a3a-4455-8758-a8d142153994)

![Screenshot_20230725-081855_Cruiser](https://github.com/devemux86/cruiser/assets/121371629/1acb5faf-5d67-4b8b-aa43-099fea9a8906)


They do not yet focus on curvy routes, but at least they are a start.

The parameters for routes around cities are still too broad for my taste. I would also like to be able to route around smaller cities, i.e. those with a population of 1000 or more.

Something like "prefer mountains" or "maximise gradients" would also be nice.
However, such an attitude should not lead to climbing a hill for 500 metres and then turning back onto the main road or driving through a small village just because it is on the hill. Such a setting only makes sense if it involves a major diversions that also includes significant gradients.

**Comment by DolomitesTom** [comment]
After several attempts, I'm afraid I have to agree with Leeb64. All the new options predominantly lead to chaotic results that are not to be expected.

**Reply by devemux86** [reply]
That's why I mentioned above that the better options are avoid road types.
(this is the method used on such related websites)

Some options may also "fight" each other. I can offer you all new BRouter options.
Then we make our tests and decide which ones we want to keep in Cruiser.

I posted the driving-cruiser profile, everyone can also test it on BRouter website.

_And better also test the BRouter bike and hiking profiles in the BRouter website._
They contain these options and are written by BRouter developers without customization.

Also keeping the discussion only here does not benefit anyone or improve profiles.
Testing the new options in the **original bike and hiking profiles** and informing BRouter developers is useful. :slightly_smiling_face:

**Comment by devemux86** [comment]
For everyone:

When reporting **BRouter** routes, please post the BRouter selected profile and options.
_We do not need Cruiser's .pref file here, but only the activated BRouter options._

Because profile tests are done on [BRouter website](https://brouter.de/brouter-web/) that offers route data table and route analysis,
to examine easily the route tags and OpenStreetMap data on the desktop.

**Reply by DolomitesTom** [reply]
To my opinion the profiles hiking and biking do not fit the needs of a motorcyclist at all.

I now have active only avoid cities, avoid motorways, avoid residential roads and award unpaved roads. Anything else leads to unwanted situations.

To my taste avoid cities could also include the avoidance of small villages as far as possible. I do not want to leave main or secondary roads just to drive 1 or 2 km through small villages except if there is a planned stop. Can you please inform Mr. Fischer about this?

Prefer forests and green areas sounds nice, but it doesn't work.

**Reply by devemux86** [reply]
> To my opinion the profiles hiking and biking do not fit the needs of a motorcyclist at all.

I only mentioned these profiles because they contain the new options and are written by the BRouter developers.

If we want report to BRouter that "Avoid cities" does not work well, we should report examples with their profiles.
So they can reproduce them. Not custom profiles that they do not have time to study and having extra changes.

> To my taste avoid cities could also include the avoidance of small villages as far as possible.

I would like that too, however in all routing engines the calculations depend on the available data.
BRouter classifies the cities with some population ranges, unfortunately not small cities / villages.

**Comment by devemux86** [comment]
If you are not satisfied with the new BRouter options we can move (some of) them in the "Advanced" group,
so that most users see just the usual avoidances with road types that we know well all this time?

**Comment by devemux86** [comment]
Work on BRouter continues and among their latest suggestions is: _"Decrease the lower population threshold to 1000"_.

- https://github.com/abrensch/brouter/pull/614

**Reply by ramondevic** [reply]
You're doing a great job with BRouter offline, too bad I can't use it because it doesn't give the street or road name information you get with GraphHopper offline :(

**Reply by umpfmpf** [reply]
Thats right! Great Job. 
I testet cruiser last weekend for 1500km against Kurviger App and Calimoto. 
And the winner is: Cruiser! With the new Brouter profile it's very close to the roads of the other programs.  It passes big cities and don't want me to turn around. I
t looks much better and i get much more information. 
Only one thing would be nice: to show every time the speedlimit.

**Comment by Leeb64** [comment]
Since nothing is happening here in terms of BRouting...
I've optimized my "temporary solution" here.
The routing prefers small streets and avoids built-up areas.
Works quite well!

[PISSESTRassen.txt](https://github.com/devemux86/cruiser/files/12505693/PISSESTRassen.txt)

**Reply by devemux86** [reply]
@Leeb64 thanks for the BRouter profile.

It is based on @poutnikl profiles?

**Reply by Leeb64** [reply]
Yes.

**Reply by poutnikl** [reply]
It is nice there are left at least links to the profile repository and wiki
what can be considered as a way of usual credit to the original source.

**Comment by Cer-o** [comment]
Voy a hacer una pregunta desde el desconocimiento absoluto de como funcionan o programan este tipo de aplicaciones.
¿No se puede copiar el algoritmo que hace a Kurviger tan especial a la hora de elegir distintos tipos de curvas, evitar centros urbanos, autovías, autopistas, peajes, etc?

**Reply by devemux86** [reply]
@Cer-o 

These routers are not open source like BRouter and the other routers.
The best thing for the community is for all of us to improve the BRouter profiles, as it works **offline**.

Estos routers no son de código abierto como BRouter y los otros routers.
Lo mejor para la comunidad es que todos mejoremos los perfiles de BRouter, ya que funciona **offline**.
